### PR TITLE
clientv3: clean up retry wrapper, remove all FailFast=false

### PR DESCRIPTION
**Pull request by gyuho**
Separate out from https://github.com/coreos/etcd/pull/8710.
c.f. https://github.com/coreos/etcd/issues/8691.

### CODE REVIEW

File: clientv3/watch.go [hunk 1]
```diff
@@ -188,7 +188,7 @@ type watcherStream struct {
 }
 
 func NewWatcher(c *Client) Watcher {
-	return NewWatchFromWatchClient(pb.NewWatchClient(c.conn))
+	return NewWatchFromWatchClient(RetryWatchClient(c))
```
Comment (xiang90) [inline clientv3/watch.go]: let us keep watch to use its own retry as it is. do not remove failfast. keep it as a todo?

**Comment by gyuho** [comment]
@xiang90 Kept `Watch` retry logic as it is. PTAL.

File: clientv3/lease.go [hunk 2]
```diff
@@ -292,9 +274,10 @@ func (l *lessor) KeepAlive(ctx context.Context, id LeaseID) (<-chan *LeaseKeepAl
 }
 
 func (l *lessor) KeepAliveOnce(ctx context.Context, id LeaseID) (*LeaseKeepAliveResponse, error) {
+	// manually retry in case "resp==nil && err==nil"
 	for {
 		resp, err := l.keepAliveOnce(ctx, id)
-		if err == nil {
+		if resp != nil && err == nil {
```
Comment (xiang90) [inline clientv3/lease.go]: why this is changed?

File: clientv3/lease.go [hunk 3]
```diff
@@ -430,10 +413,9 @@ func (l *lessor) recvKeepAliveLoop() (gerr error) {
 			if canceledByCaller(l.stopCtx, err) {
 				return err
 			}
-		} else {
+		} else if stream != nil {
```
Comment (xiang90) [inline clientv3/lease.go]: why this is changed?

File: clientv3/lease.go [hunk 4]
```diff
@@ -461,8 +443,8 @@ func (l *lessor) recvKeepAliveLoop() (gerr error) {
 // resetRecv opens a new lease stream and starts sending keep alive requests.
 func (l *lessor) resetRecv() (pb.Lease_LeaseKeepAliveClient, error) {
 	sctx, cancel := context.WithCancel(l.stopCtx)
-	stream, err := l.remote.LeaseKeepAlive(sctx, grpc.FailFast(false))
-	if err != nil {
+	stream, err := l.remote.LeaseKeepAlive(sctx)
+	if stream == nil || err != nil {
```
Comment (xiang90) [inline clientv3/lease.go]: why this is changed?

File: clientv3/lease.go (hunk 3 above)
Comment (gyuho) [reply clientv3/lease.go]: I was following the pattern here

```go
ws, err = w.remote.Watch(w.ctx, grpc.FailFast(false)); ws != nil && err == nil {
```

where it's possible `ws==nil && err == nil`.

Do you remember why we had this in the first place? In other retry PR, I have tests failing (`nil pointer`) from `stream == nil && err == nil`. Will investigate more.

Comment (xiang90) [reply clientv3/lease.go]: i do not want to mix fixes together into one commit. this pr should only focus on removing the failfast option and use the shared retry func.

Comment (gyuho) [reply clientv3/lease.go]: Agree. Will revert this change.

**Comment by gyuho** [comment]
@xiang90 PTAL.

File: clientv3/kv.go [hunk 5]
```diff
@@ -90,15 +88,15 @@ func (resp *TxnResponse) OpResponse() OpResponse {
 }
 
 type kv struct {
-	remote pb.KVClient
+	readWrite pb.KVClient
```
Comment (xiang90) [inline clientv3/kv.go]: revert this?

File: clientv3/kv.go [hunk 6]
```diff
@@ -132,53 +130,37 @@ func (kv *kv) Txn(ctx context.Context) Txn {
...
 		r := &pb.DeleteRangeRequest{Key: op.key, RangeEnd: op.end, PrevKv: op.prevKV}
-		resp, err = kv.remote.DeleteRange(ctx, r)
+		resp, err = kv.readWrite.DeleteRange(ctx, r)
 		if err == nil {
 			return OpResponse{del: (*DeleteResponse)(resp)}, nil
 		}
 	case tTxn:
 		var resp *pb.TxnResponse
-		resp, err = kv.remote.Txn(ctx, op.toTxnRequest())
+		// op.isWrite()==false, then readOnly.Txn
```
Comment (xiang90) [inline clientv3/kv.go]: revert?

File: clientv3/retry.go [hunk 7]
```diff
@@ -268,6 +268,80 @@ func (rcc *nonRepeatableClusterClient) MemberUpdate(ctx context.Context, in *pb.
...
+
+func (rmc *nonRepeatableMaintenanceClient) Defragment(ctx context.Context, in *pb.DefragmentRequest, opts ...grpc.CallOption) (resp *pb.DefragmentResponse, err error) {
+	err = rmc.nonRepeatableRetry(ctx, func(rctx context.Context) error {
+		resp, err = rmc.mc.Defragment(rctx, in, opts...)
+		return err
+	})
+	return resp, err
+}
+
+func (rmc *nonRepeatableMaintenanceClient) MoveLeader(ctx context.Context, in *pb.MoveLeaderRequest, opts ...grpc.CallOption) (resp *pb.MoveLeaderResponse, err error) {
```
Comment (xiang90) [inline clientv3/retry.go]: this is repeatable.

File: clientv3/retry.go [hunk 8]
```diff
@@ -101,63 +101,63 @@ func (c *Client) newAuthRetryWrapper() retryRPCFunc {
 	}
 }
 
-// RetryKVClient implements a KVClient that uses the client's FailFast retry policy.
+// RetryKVClient implements a KVClient.
 func RetryKVClient(c *Client) pb.KVClient {
-	readRetry := c.newRetryWrapper(isReadStopError)
-	writeRetry := c.newRetryWrapper(isWriteStopError)
+	repeatableRetry := c.newRetryWrapper(isReadStopError)
```
Comment (xiang90) [inline clientv3/retry.go]: rename isReadStopError to isRepeatableStopError. same for nonrepeatable.

File: clientv3/retry.go [hunk 9]
```diff
@@ -101,63 +101,63 @@ func (c *Client) newAuthRetryWrapper() retryRPCFunc {
...
-		&retryWriteKVClient{retryBasic, retryAuthWrapper},
+	return &repeatableKVClient{
+		&nonRepeatableKVClient{retryBasic, retryAuthWrapper},
 		retryAuthWrapper}
 }
 
-type retryKVClient struct {
-	*retryWriteKVClient
-	readRetry retryRPCFunc
+type repeatableKVClient struct {
```
Comment (xiang90) [inline clientv3/retry.go]: retryKVClient.

File: clientv3/txn.go [hunk 10]
```diff
@@ -136,30 +134,15 @@ func (txn *txn) Else(ops ...Op) Txn {
...
 
-func (txn *txn) commit() (*TxnResponse, error) {
 	r := &pb.TxnRequest{Compare: txn.cmps, Success: txn.sus, Failure: txn.fas}
 
-	var opts []grpc.CallOption
-	if !txn.isWrite {
-		opts = []grpc.CallOption{grpc.FailFast(false)}
-	}
-	resp, err := txn.kv.remote.Txn(txn.ctx, r, opts...)
+	// TODO: txn.isWrite==false, then readOnly.Txn
```
Comment (xiang90) [inline clientv3/txn.go]: revert

File: clientv3/retry.go [hunk 11]
```diff
@@ -101,63 +101,63 @@ func (c *Client) newAuthRetryWrapper() retryRPCFunc {
...
 		return err
 	})
 	return resp, err
 }
 
-func (rkv *retryWriteKVClient) Txn(ctx context.Context, in *pb.TxnRequest, opts ...grpc.CallOption) (resp *pb.TxnResponse, err error) {
-	err = rkv.writeRetry(ctx, func(rctx context.Context) error {
-		resp, err = rkv.KVClient.Txn(rctx, in, opts...)
+func (rkv *nonRepeatableKVClient) Txn(ctx context.Context, in *pb.TxnRequest, opts ...grpc.CallOption) (resp *pb.TxnResponse, err error) {
+	err = rkv.nonRepeatableRetry(ctx, func(rctx context.Context) error {
```
Comment (xiang90) [inline clientv3/retry.go]: add a todo here: if txn only contains read operation, it can be repeatable.

File: clientv3/retry.go [hunk 12]
```diff
@@ -218,35 +218,51 @@ func (rlc *retryLeaseClient) LeaseKeepAlive(ctx context.Context, opts ...grpc.Ca
...
+	repeatableRetry := c.newRetryWrapper(isReadStopError)
+	nonRepeatableRetry := c.newRetryWrapper(isWriteStopError)
+	cc := pb.NewClusterClient(c.conn)
+	return &repeatableClusterClient{&nonRepeatableClusterClient{cc, nonRepeatableRetry}, repeatableRetry}
 }
 
-// RetryClusterClient implements a ClusterClient that uses the client's FailFast retry policy.
-func RetryClusterClient(c *Client) pb.ClusterClient {
-	return &retryClusterClient{pb.NewClusterClient(c.conn), c.newRetryWrapper(isWriteStopError)}
+type repeatableClusterClient struct {
```
Comment (xiang90) [inline clientv3/retry.go]: retryClusterClient.

File: clientv3/retry.go [hunk 13]
```diff
@@ -268,6 +268,80 @@ func (rcc *nonRepeatableClusterClient) MemberUpdate(ctx context.Context, in *pb.
...
 
+// RetryMaintenanceClient implements a Maintenance.
+func RetryMaintenanceClient(c *Client, conn *grpc.ClientConn) pb.MaintenanceClient {
+	repeatableRetry := c.newRetryWrapper(isReadStopError)
+	nonRepeatableRetry := c.newRetryWrapper(isWriteStopError)
+	mc := pb.NewMaintenanceClient(conn)
+	return &repeatableMaintenanceClient{&nonRepeatableMaintenanceClient{mc, nonRepeatableRetry}, repeatableRetry}
+}
+
+type repeatableMaintenanceClient struct {
```
Comment (xiang90) [inline clientv3/retry.go]: retry...Client

File: clientv3/retry.go [hunk 14]
```diff
@@ -342,99 +342,147 @@ func (rmc *nonRepeatableMaintenanceClient) MoveLeader(ctx context.Context, in *p
 	return resp, err
 }
 
-type retryAuthClient struct {
-	pb.AuthClient
-	writeRetry retryRPCFunc
+type repeatableAuthClient struct {
```
Comment (xiang90) [inline clientv3/retry.go]: retryAuthClient

File: clientv3/retry.go [hunk 15]
```diff
@@ -163,37 +163,37 @@ func (rkv *nonRepeatableKVClient) Compact(ctx context.Context, in *pb.Compaction
 	return resp, err
 }
 
-type retryLeaseClient struct {
+type repeatableLeaseClient struct {
```
Comment (xiang90) [inline clientv3/retry.go]: retryLeaseClient.

**Comment by gyuho** [comment]
@xiang90 PTAL. Kept the old `struct` since it's easier to wrap auth wrapper.

**Comment by xiang90** [comment]
lgtm

**Pull request merged**